core/storage/statistics/gae_models.py

```diff
@@ -587,114 +590,219 @@ def get_versions(cls, exploration_id):
+        # simply be ignored in the result of this function. It will be included
+        # during the next call.
+        main_shard = cls._get_model(
+            exploration_id, exploration_version, state_name, 0)
+
+        if main_shard:
+            all_models = [main_shard]
+            if main_shard.shard_count > 0:
+                shard_ids = [
+                    cls._get_entity_id(
+                        exploration_id, exploration_version, state_name,
+                        shard_id)
+                    for shard_id in xrange(1, main_shard.shard_count + 1)]
+                all_models += cls.get_multi(shard_ids)
+            return all_models
+        else:
+            return None
 
     @classmethod
-    def create_or_update(
+    def _insert_submitted_answers_unsafe(
             cls, exploration_id, exploration_version, state_name,
-            interaction_id, answers_list):
-        entity_id = cls._get_entity_id(
-            exploration_id, str(exploration_version), state_name)
-        instance = cls(id=entity_id, exploration_id=exploration_id,
-                       exploration_version=exploration_version,
-                       state_name=state_name,
-                       interaction_id=interaction_id,
-                       answers_list=answers_list)
-        return instance
+            interaction_id, new_submitted_answer_dict_list):
+        # The main shard always needs to be retrieved. At most one other shard
+        # needs to be retrieved (the last one).
+        main_shard = cls._get_model(
+            exploration_id, exploration_version, state_name, 0)
+        last_shard = main_shard
+
+        if not main_shard:
+            entity_id = cls._get_entity_id(
+                exploration_id, exploration_version, state_name, 0)
+            main_shard = cls(
+                id=entity_id, exploration_id=exploration_id,
+                exploration_version=exploration_version, state_name=state_name,
+                shard_id=0, interaction_id=interaction_id, shard_count=0,
+                submitted_answer_list=[])
+            last_shard = main_shard
+        elif main_shard.shard_count > 0:
+            last_shard = cls._get_model(
+                exploration_id, exploration_version, state_name,
+                main_shard.shard_count)
+
+        sharded_answer_lists, sharded_answer_list_sizes = cls._shard_answers(
+            last_shard.submitted_answer_list,
+            last_shard.accumulated_answer_json_size,
+            new_submitted_answer_dict_list)
+        new_shard_count = main_shard.shard_count + len(sharded_answer_lists) - 1
+
+        # Collect all entities to update to efficiently send them as a single
+        # update.
+        entities_to_put = []
+        last_shard_is_main = main_shard.shard_count == 0
+
+        # Update the last shard if it changed.
+        if sharded_answer_lists[0] != last_shard.submitted_answer_list:
```

Comment:
Good idea to use size; done. It wouldn't change in the case when adding a new answer requires adding a new shard without changing the last one.